neps/nep-0488.md

```diff
@@ -0,0 +1,1537 @@
+
+- $x \in F_{p^2}$ `[u8; 96]`:
+  - $c_1 \in F_p$ `[u8; 48]`
+  - $c_0 \in F_p$ `[u8; 48]`
+- $y \in F_{p^2}$ `[u8; 96]`:
+  - $c_1 \in F_p$ `[u8; 48]`
+  - $c_0 \in F_p$ `[u8; 48]`
+
+*The second bit* is used to mark a point on infinity, if the second bit is set to 1 — it is an infinity point.
+
+Encoding point on infinity:
+
+```bash
+let x: [u8; 192] = [0; 192];
+x[0] = x[0] | 0x40;
+```
+
+The rule of encoding is consistent with zkcrypto[^53] and with implementation in milagro lib[^29].
+
+#### Compressed points on twisted curve E'(Fp2)
+
+The points on the curve represent by affine coordinates: `(x: Fp2, y: Fp2)`. The elements from $E'(F_{p^2})$ in compressed form are encoded in `[u8; 96]` as *big-endian* encoded $x \in F_{p^2}$. The $y$  coordinate can be detected by the formula: $y = \pm \sqrt{x^3 + 4}$.
+
+*The first bit* should be set as 1. This bit indicates that point is encoded in compressed form.
+
+*The second bit* is used to mark a point on infinity, if the second bit is set to 1 — it is an infinity point.
+
+To represent the sign of the $y$ *the third bit* of x encoding is used. If the first bit = 0 the $y$ is positive, if the first bit = 1 the $y$ is negative. We will consider the number positive the smallest between $y$  and $-y$: first compare $c_1$ and second $c_0$.
+
+The $x \in F_{p^2}$ encoded as `[u8; 96]` bytes according to the rules from section  “Extension fields elements $F_{p^2}$” .
+
+The point on $E'(F_{p^2})$ with negative $y$ coordinate encoded as `[u8; 96]` bytes:
+
+```rust
+let x: [u8; 96] = encodeFp2(x);
+x[0] = x[0] | 0x80;
+x[0] = x[0] | 0x20;
+```
+
+Encoding point of infinity:
+
+```rust
+let x: [u8; 96] = [0; 96];
+x[0] = x[0] | 0x80;
+x[0] = x[0] | 0x40;
+```
+
+The rule of encoding is consistent with zkcrypto[^53] and with implementation in milagro lib[^29].
+
+### Host functions
+
+#### bls12381_g1_sum
+
+***Description:***
+
+The function computes the sum of the signed elements of the BLS12-381 curve. The input is an arbitrary number of pairs $(p_i, s_i)$, where $p_i \in E(F_p)$ is point on elliptic curve and $s_i \in \textbraceleft 0, 1 \textbraceright$ is the point sign. The output is one point from $E(F_p)$ equal to $\sum (-1)^{s_i}p_i$.
+
+The $E(F_p)$ curve,  points on the curve, multiplication on -1,  and the addition operation are defined in the BLS12-381 Curve Specification section.
+
+Note: we take as input any points on the curve, not only from   $G_1$
+
+***Input:*** the sequence of pairs $(p_i, s_i)$, where $p_i \in E(F_p)$ is point and $s_i \in \textbraceleft 0, 1 \textbraceright$ is sign, each point encoded in decompress form as $(x\colon F_p, y\colon F_p)$ and sign encoded in one byte with only two allowed values: 0, 1. Expected `97*k` bytes as an input that is interpreted as byte concatenation of `k` slices, each slice is the uncompressed point from $E(F_p)$ and a bool value for point sign. More details are in the Curve Points Encoding section.
```

Comment:
Yes, I have done it in the same way as in bn254 implementation. Also in our multiexp functions, scalars go after points. I just try to make it similar. If someone wants to switch from bn254 to bls12-381 it can be convenient to have the same ABI. 

I don't have a strong opinion about what is the right way. From the implementation point of view, it is no different.

Reply:
Uhhh… in my copy of logic.rs, I see ``value` - sequence of (sign:bool, g1:G1), where [...]` as a comment on `alt_bn128_g1_sum`, so the sign is before. Am I missing something?

Reply:
Hmm... I see, it is not consistent with https://github.com/near/NEPs/issues/98.

Ok, I will change it. Thanks for noticing it.

Reply:
fixed: https://github.com/near/NEPs/pull/488/commits/d9b039a05f6552e8005ebd4102c032def6cf4226